Sources/SwiftFormatRules/DoNotUseSemicolons.swift

```diff
@@ -22,52 +22,73 @@ import SwiftSyntax
+  private func nodeByRemovingSemicolons<
+    ItemType: Syntax & SemicolonSyntax & Equatable, NodeType: SyntaxCollection
+  >(from node: NodeType, nodeCreator: ([ItemType]) -> NodeType) -> NodeType
+  where NodeType.Element == ItemType {
+    var newItems = Array(node)
 
     // Because newlines belong to the _first_ token on the new line, if we remove a semicolon, we
     // need to keep track of the fact that the next statement needs a new line.
     var previousHadSemicolon = false
-    for (idx, item) in items.enumerated() {
+    for (idx, item) in node.enumerated() {
 
       // Store the previous statement's semicolon-ness.
       defer { previousHadSemicolon = item.semicolon != nil }
 
+      // Check for semicolons in statements inside of the item, because code blocks may be nested
+      // inside of other code blocks.
+      guard let visitedItem = visit(item) as? ItemType else {
+        return node
+      }
+
       // Check if we need to make any modifications (removing semicolon/adding newlines)
-      guard item.semicolon != nil || previousHadSemicolon else {
+      guard visitedItem != item || item.semicolon != nil || previousHadSemicolon else {
         continue
       }
 
-      var newItem = item
+      var newItem = visitedItem
 
-      if previousHadSemicolon {
-        // Ensure the leading trivia for this statement has a newline.
-        let firstTok = item.firstToken
+      // Check if the leading trivia for this statement needs a new line.
+      if previousHadSemicolon, let firstToken = newItem.firstToken,
+        !firstToken.leadingTrivia.containsNewlines
+      {
+        let leadingTrivia = .newlines(1) + firstToken.leadingTrivia
         newItem = replaceTrivia(
-          on: item,
-          token: firstTok,
-          leadingTrivia: firstTok?.leadingTrivia.withOneLeadingNewline() ?? .newlines(1)
-        ) as! CodeBlockItemSyntax
+          on: newItem,
+          token: firstToken,
+          leadingTrivia: leadingTrivia
+        ) as! ItemType
       }
 
       // If there's a semicolon, diagnose and remove it.
-      if idx < items.count {
-        diagnose(.removeSemicolonAndMove, on: item)
-      } else {
-        diagnose(.removeSemicolon, on: item)
+      if let semicolon = item.semicolon {
+        newItem = newItem.withSemicolon(nil)
+        if idx < node.count - 1 {
+          diagnose(.removeSemicolonAndMove, on: semicolon)
+        } else {
+          diagnose(.removeSemicolon, on: semicolon)
+        }
       }
-      newItem = newItem.withSemicolon(nil)
       newItems[idx] = newItem
     }
-    return SyntaxFactory.makeCodeBlockItemList(newItems)
+    return nodeCreator(newItems)
   }
 
-  public override func visit(_ node: CodeBlockSyntax) -> Syntax {
-    return node.withStatements(transformItems(node.statements))
+  public override func visit(_ node: CodeBlockItemListSyntax) -> Syntax {
+    return nodeByRemovingSemicolons(from: node) { SyntaxFactory.makeCodeBlockItemList($0) }
```

Comment:
It's equivalent, but when you have a closure where you're just passing through the arguments to another function in the same order, you can usually pass a reference to that function directly, so this should also work:

```swift
nodeByRemovingSemicolons(from: node, nodeCreator: SyntaxFactory.makeCodeBlockItemList)
```

My personal style is to prefer function references instead of creating new closure literals when possible, but it's not a *strong* preference and I've also never bothered looking into whether there were particular advantages to one or the other in terms of compiler performance.

Reply:
I forgot about that! Avoiding unnecessary closures seems like a win to me. Changed.